README.md
# MASM64 with RadASM2

This project consists of related and needed files for using MASM64 SDK with RadASM2.

![](https://img.shields.io/badge/Assembler-MASM64%20v14.xx-green.svg?style=flat-square&logo=visual-studio-code&logoColor=white&colorB=1CC887) [![](https://img.shields.io/badge/RadASM%20-v2.2.2.x%20-red.svg?style=flat-square&colorB=C94C1E&logo=data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABQAAAAgCAYAAAASYli2AAACcklEQVR42tWVXWiPURzHz/FyQZOiVuatuFEoKzfKSCs35EJeCqFcEEa5s2heNrXiApuXFDYveUlKSywlIRfczM0WjZvJlGKTRLb5fHvOU6fT+T/PY3bj1Kff8z8vn+f8znPO+dshihnBYv8L4awRcl2FRTarBy8bQzgEjdbabzl9nxCW2IwOFYTrsBTKEH7PET4lLLYlGpcTrkC5qxqL8HeO8CVhoQ0qRxMOw34Y5TVVIPyYI+whTLVehZ9iWgZAL1mN8G6GbArhA/TZEilqKx2HCbADXkAV0oESwhOEfdChbXOUh1ovxS+wlcH3aNvC82VX3wx7Qyl9NhEugXZEU7ixX8E6Br13nTVDPU927R3QCl0wTX2h2rUNQqUv/ATLkHUGM1hLuBF8pFipZ+zBcIZKpw1O0vjYk24mnIXxEZHGNMIBxgxJ2M2P2PF7DafhGh1/0G8Gzzv1cWASfIZn0EJ7VzpIQqWyUguulFUXiDXwApxhYE9O2ibc2PMJNbAxkp5Oyh3NGvHzQkJPrK/aANtLjNNuOAU3kf/KFTrpGsJtaIdxbu3C0gvn4Dzi3qLCI3Su4/cCnnfDBvcCv/yEW0a7o6gwWI5tJvniMwutYZbQa9elsUqzgun/JKStjKAzvAvmDXuG1M1xqerkTAyG6Cy3FREeM8k2kag6MomvcBGaefG7LOF6k1wK6SUbFl0iOpqt/v+NjYjmEva4NQpPi9K6b5JN/UiXQTg+vbF1nlc4USytPpNcok1Iuk1G0eWgS0Hnd3akXbeIbuqWvP9lXxhOW2k9cOvzMJZWUWG/Sf4/lNbbv5GEwjeSSIaof7iitPwBoSgbVud1Jo0AAAAASUVORK5CYII=)](http://www.softpedia.com/get/Programming/File-Editors/RadASM.shtml)

## Setup
* Extract [RadASM2][web-github-radasm2-release] to wherever you want.
* Unzip [MASM64 SDK Package][web-masm64-sdk] under `C:\`
* Unzip this project inside your RadASM2 folder.
* Open `RadASM.ini` with your fav editor and add `MASM64` under `Assembler` section; look at the sample below:
```
[Assembler]
Assembler=MASM64,masm,UASM32,UASM64,nasm,mingw,MINGW32,MINGW64
```
**NOTE**: You don't have to write that much programming languages like me. But you must add at least one programming language to run RadASM2.


## Notes
For a seamless work via **MASM64 with RadASM2**, you must be sure that you've installed **MASM64 SDK** properly. 

Your folder structure should look like below:
```
C:\masm64
C:\masm64\bin64
C:\masm64\include64
C:\masm64\lib64
C:\masm64\macros64
```

Creators of MASM64 SDK cannot distribute Microsoft binaries; me neither. So you have to install a Windows SDK to your machine and copy some binaries under `c:\masm64\bin64` and `c:\masm64\bin64\1033`

Inside [MASM64 SDK archive][web-masm64-sdk] there is a `bin64\FileList.txt` file specifies which binaries should be istalled to where:
```
**********************************************
Files that must be obtained from Microsoft
Usually this is from a Visual Studio download.
**********************************************

------------------------------------------------------
This is a link to the community edition of build tools
------------------------------------------------------
https://visualstudio.microsoft.com/downloads/?q=build+tools

-----------------
Files in => bin64
-----------------
cvtres.exe
dumpbin.exe
editbin.exe
lib.exe
link.exe
link.exe.config
ml64.exe
msobj140.dll
mspdb140.dll
mspdbcmf.exe
mspdbcore.dll
mspdbsrv.exe
mspdbst.dll
mspft140.dll
msvcdis140.dll
msvcp140.dll
nmake.exe
tbbmalloc.dll

----------------------
Files in => bin64\1033
----------------------
atlprovui.dll
bscmakeui.dll
clui.dll
cvtresui.dll
linkui.dll
LocalESPCui.dll
mspdbcmfui.dll
mspft140ui.dll
nmakeui.dll
pgort140ui.dll
pgoui.dll
```

## TODO

* [x] Add/Edit Common Control messages and constants
* [x] Add Console64 App Build.bat
* [x] Fix/Clean build parameters
* [x] Add batch files which can handle projects with/without rc files
* [x] Add templates
  * [x] Add Console64 templates
  * [x] Add DLL64 templates
  * [x] Add Lib64 templates

## Resources
* [MASM64 Beta Version 2 Forum Topic][web-masm64-forum-topic]
* [MASM64 Beta Version 2 Download Link][web-masm64-sdk]
* [Visual Studio Build Tools][web-ms-vs-build-tools]
* [MASM32 SDK][web-masm32-sdk]
* [RadASM2][web-github-radasm2]
* [RadASM2 v2.2.2.4][web-github-radasm2-release]

## Thanks
I have inspired from [mrfearless][web-github-fearless] while creating this repo. He has already maintained a repo named [UASM-with-RadASM2][web-github-uasm-with-radasm2]. [UASM][web-uasm] is MASM compatible open source assembler project. Also check that out!

## Contributors
* [Blue DeviL // SCT][web-github-bluedevil]

## License
This project is under GPLv3 license.

[web-github-bluedevil]: https://github.com/blue-devil
[web-win-sdk-archive]: https://developer.microsoft.com/en-us/windows/downloads/sdk-archive
[web-masm32-sdk]: http://www.masm32.com/download.htm
[web-masm64-sdk]: http://www.masm32.com/download/masm64.zip
[web-masm64-forum-topic]: https://masm32.com/board/index.php?topic=10052.0
[web-ms-vs-build-tools]: https://visualstudio.microsoft.com/downloads/?q=build+tools
[web-github-radasm2-release]: https://github.com/blue-devil/RadASM2/blob/master/Release/RadASM-2.2.2.4-FullPackage.zip
[web-github-radasm2]: https://github.com/mrfearless/RadASM2
[web-github-fearless]: https://github.com/mrfearless
[web-github-uasm-with-radasm2]: https://github.com/mrfearless/UASM-with-RadASM
[web-uasm]: http://www.terraspace.co.uk/uasm.html

CHANGELOG.md
# Changelog

## [2022.10.17] - 17-10-2022
### Changes
* `MASM64Stdlib.api` deleted, this api file was an old file
* UPDATE: api files fix typo and cosmetic update
* UPDATE: api files

## [2022.10.13] - 13-10-2022
### Added
* Added Base WndProc Sniplet
* OPENFILENAMEW added to MASM64ApiStruct.api
* Some interger types from Windows base types added to MASM64.ini:
  * INT, INT8, INT16, INT32, INT64, INT_PTR
  * UINT8, UINT16, UINT32, UINT64, UINT_PTR
* `MapViewOfFileExNuma` prototype added to MASM64ApiCall.api
* `MapViewOfFileExNuma`, `MapViewOfFileEx`, `MapViewOfFile` functions' `dwDesiredAccess` parameters added to `MASM64ApiConst.api`
* New winapi function protoypes added to `MASM64ApiCall.api`
  * CreateFileMappingNumaW, CreateFileMappingNumaA, CreateFileMappingNuma
  * CreateFileMappingW, CreateFileMappingA
* `flProtect` parameters added:
  * PAGE_EXECUTE_READ, PAGE_EXECUTE_READWRITE, PAGE_EXECUTE_WRITECOPY, PAGE_READONLY, PAGE_READWRITE, PAGE_WRITECOPY
  * SEC_COMMIT,SEC_IMAGE,SEC_IMAGE_NO_EXECUTE,SEC_LARGE_PAGES,SEC_NOCACHE,SEC_RESERVE,SEC_WRITECOMBINE
  * NUMA_NO_PREFERRED_NODE
* OPENFILENAME struc's flags added to `MASM64ApiWord.api`
  * OFN_*
* WM_SIZE message wParam parameters added to `MASM64ApiWord.api`
* STARTUPINFOW, STARTUPINFOEX{A/W}, added to `MASM64ApiWord.api`
  * `STARTF_*` flag parameters added to MASM64ApiWord.api
* GetStartupInfoA and GetStartupInfoW, CreateProcessA and CreateProcessW added to `MASM64ApiCall.api`
* **Process Creation Flags** added to `MASM64ApiWord.api`
* DialogWithManifest sample project added
* ADDED: TUTE14 - Process
* ADDED: LoadImage, CreateThread and CreateFont Constants
* FIX: LoadCursor and LoadIcon apis shouldn't get hInst as parameter. They should get null.
* FIX: Some apis mistakenly get 0 instead of null
* FIX: Some cosmetic changes has been made
* ADDED: two new code sniplet added: DialogBoxParam and CreateDialogParam
* ADDED: TUTE15
* ADDED: TUTE16
* ADDED: `BOOL`, `BOOLEAN` to MASM64.ini
* ADDED: DLL reason flags to api files
* ADDED: TUTE17
* ADDED: Common Control style constants to api files
* ADDED: TUTE 18
* ADDED: MASM64StructConst.api (not working rn)
* ADDED: MASM64ApiMacroCall.api
* ADDED: HDC and HBITMAP to MASM64.ini for color hiliting
* ADDED: TUTE20, TUTE21, TUTE22
* Added: SetWindowHookEx first parameters added to MASM64ApiConst.api
* Added: Hook Codes added to MASM64ApiWord.api
* Added: GetClassLongPtr and SetClassLongPtr 2nd parameteres to MASM64ApiConst.api
* Added: SetClassLongPtr, GetClassLongPtr to MASM64ApiCall.api
* ADDED: ULONG,ULONGLONG,ULONG_PTR,ULONG32,ULONG64 to MASM64.ini
* ADDED: MSLLHOOKSTRUCT and constants
* ADDED: NOTIFYICONDATA and constants
* ADDED: TUTE28
* ADDED: DuplicateHandle's constants
* ADDED: Status Flags
* ADDED: Compatibility macros
* ADDED: DEBUG_EVENT Structure Constants
* ADDED: MessageBox constants
* ADDED: Lots of new constants and macros of window controls
* ADDED: TUTE32
* ADDED: Batch file for building LIB files
* ADDED: TUTE33 but not finished
* ADDED: TUTE34
* ADDED:Adding and updating Common Control messages and constants finished


### Changes
* Update: version
* In Win64 3 parameters added to OPENFILENAME: pvReserved,dwReserved,FlagsEx. Those parameters added to OPENFILENAME and OPENFILENAMEA
* FIX: Taskbar constants edited
* UPDATE: MASM64ApiWord.api common controls updated
* UPDATE: Cosmetic
* FIX: Wrong LIB64 parameters fixed in MASM64.ini
* FIX: WindowFromPoint proto fixed in MASM64ApiCall.api
* FIX: added SetClassLong 2nd parameter
* FIX: TUTE24
* UPDATE: MASM64.ini Windows DataTypes are updated
* UPDATE: According to [Directives Reference](https://learn.microsoft.com/en-us/cpp/assembler/masm/directives-reference?view=msvc-170) MASM64.ini updated
* UPDATE: [Symbols reference](https://learn.microsoft.com/en-us/cpp/assembler/masm/symbols-reference?view=msvc-170) checked and MASM64.ini updated
* UPDATE: MessageBooep constants updated
* UPDATE: TUTE28 - [Accessing Union elements inside a Structure in MASM64](https://masm32.com/board/index.php?topic=10387.0)
* UPDATE: EM_REDO was missing on sendmessage MASM64ApiConst.api
* UPDATE: TUTE33
* UPDATE: cosmetic on projects directory

## [2022.09.01] - 01-09-2022
### Added
* New data types added to MASM64Type.api
* Modal and Modeless DialogApp Templates added.
* RGB macro snippet added
* Some handle values added to MASM64.ini (HICON, HCURSOR, HMENU)

### Changes
* Using [Windows Data Types][web-msdocs-windows-data-types] all data types inside MASM64Type.api checked and fixed
* Window Class Styles moved from MASM64Message.api to MASM64ApiWord.api

## [2022.08.25] - 25-08-2022
### Added
* .SKIP .UNTILCXZD .UNTILCXZE .UNTILCXZNE .UNTILCXZED .UNTILCXZNED .ENDF keyword added to MASM64.ini for keyword highlighting
* Window class styles added to MASM64Message.api. Three of them deprecated and one is new. ([Source][web-msdocs-windows-class-styles])
  * CS_VREDRAW,equ 1h
  * CS_HREDRAW,equ 2h
  * CS_KEYCVTWINDOW,equ 4h      // deprecated?
  * CS_DBLCLKS,equ 8h
  * CS_OWNDC,equ 20h
  * CS_CLASSDC,equ 40h
  * CS_PARENTDC,equ 80h
  * CS_NOKEYCVT,equ 100h        // deprecated?
  * CS_NOCLOSE,equ 200h
  * CS_SAVEBITS,equ 800h
  * CS_BYTEALIGNCLIENT,equ 1000h
  * CS_BYTEALIGNWINDOW,equ 2000h
  * CS_PUBLICCLASS,equ 4000h    // deprecated?
  * CS_GLOBALCLASS,equ CS_PUBLICCLASS
  * CS_IME,equ 10000h
  * CS_DROPSHADOW,equ 20000h    // new
* CW_USEDEFAULT value added to MASM64Message.api
  * CW_USEDEFAULT,equ 80000000h


[web-msdocs-windows-data-types]: https://docs.microsoft.com/en-us/windows/win32/winprog/windows-data-types
[web-msdocs-windows-class-styles]: https://docs.microsoft.com/en-us/windows/win32/winmsg/window-class-styles

MASM64/Projects/Iczelion/TUTE13/TUTE13.Asm
;  _______________________________________________________________________
; |  ______ __           ,.'   __.'.__  .   ______               ___      |
; | |      |  |          /b`  '-.   .-'  d\|      ;             |   |     |
; | |   __ |  |         .5h     /.'.\    =5|   _   \            |   |     |
; | |.  1 \|  |         (0=h    '   '    =C)   1\   \        .--:.  |     |
; | |   1_/|  |         ?3==,          .==7|   | \   \       |__|   |     |
; | |.     |  .-- -.-----{=`==oo____oo==`=}|   |  |.----.-- -.--:.  |     |
; | |.  __ |  |  V |  ^__|'"o58888888888,; |.  |  ||  ^_|  V |  |.  |     |
; | |:  1 \|__|____|_____| `?88P^\,?88^\P  |.  | / |____|\___|__|:  |___  |
; | |:  1_/   /             C8?\__d88\_/'  |:  1/   / .         |:  1   | |
; | |::.. .  /              `8o8888/\88P   |::.. . /            |::.. . | |
; | `-------'                '7oo88oo8P    `------'             `-------' |
; |                      Cooo##~\/\/\/~====>        github.com/blue-devil |
; |                                                  gitlab.com/bluedevil |
; | Freedom doesn't come from second-hand thoughts.       www.sctzine.com |
; |_______________________________________________________________________|
; | @author   : Blue DeviL <bluedevil@sctzine.com>                        |
; | @tester   : ErrorInside <errorinside@sctzine.com>                     |
; | @IDE      : RadASM v2.2.2.4                                           |
; | @template : Blue DeviL                                                |
; | @date     : 08/09/2022                                                |
; | @license  : n/a                                                       |
; |_______________________________________________________________________|
; |                                                                       |
; |           Iczelion's Win32 Assembly Tutorials with MASM64             |
; |                  Tutorial 13 - Memory Mapped Files                    |
; \_______________________________________________________________________/


    OPTION DOTNAME                          ; required for macro files
    option casemap:none                     ; case sensitive

; _________________________________________________________________________
; MASM64 macros

    include \masm64\include64\win64.inc     ; main include file
    include \masm64\macros64\vasily.inc     ; main macro file
    include \masm64\macros64\macros64.inc   ; auxillary macro file

    STACKFRAME                              ; create a default stack frame

; _________________________________________________________________________
; include files

    include \masm64\include64\kernel32.inc
    include \masm64\include64\user32.inc
    include \masm64\include64\comdlg32.inc

; _________________________________________________________________________
; libraries

    includelib \masm64\lib64\kernel32.lib
    includelib \masm64\lib64\user32.lib
    includelib \masm64\lib64\comdlg32.lib

; _________________________________________________________________________
; funtion prototypes

    WndProc             PROTO   :HWND,:UINT,:WPARAM,:LPARAM
    CloseMapFile        PROTO

; _________________________________________________________________________
; constant variables

.const
    IDM_OPEN        equ 1
    IDM_SAVE        equ 2
    IDM_EXIT        equ 3
    MAXSIZE         equ 260
; _________________________________________________________________________
; initialized variables

.data
    ClassName       db "Win64ASMFileMappingClass",0
    AppName         db "Win64 ASM File Mapping Example",0
    MenuName        db "FirstMenu",0
    ofn             OPENFILENAME <>
    FilterStr       db "All Files",0,"*.*",0
                    db "Text Files",0,"*.txt",0,0
    buffer          db MAXSIZE dup(0)
    hMapFile        HANDLE 0

; _________________________________________________________________________
; uninitialized variables

.data?
    CommandLine     LPSTR ?
    hInstance       HINSTANCE ?
    hIcon           HICON ?
    hCursor         HCURSOR ?
    hFileRead       HANDLE ?
    hFileWrite      HANDLE ?
    hMenu           HANDLE ?
    pMemory         QWORD ?
    SizeWritten     QWORD ?

.code

WinMainCRTStartup proc
    
    invoke  GetModuleHandle, NULL
    mov     hInstance, rax
    
    invoke  GetCommandLine
    mov     CommandLine, rax
    
    invoke  WinMain, hInstance, NULL, CommandLine, SW_SHOWDEFAULT
    invoke  ExitProcess, eax
    
    ret

WinMainCRTStartup endp

WinMain proc hInst:HINSTANCE,hPrevInst:HINSTANCE,CmdLine:LPSTR,CmdShow:DWORD
    
    LOCAL   wcex:WNDCLASSEX
    LOCAL   msg:MSG
    LOCAL   hWnd:HWND
    
    invoke  LoadIcon, NULL, IDI_APPLICATION
    mov     hIcon, rax
    invoke  LoadCursor, NULL, IDC_ARROW
    mov     hCursor, rax
    
    mov     wcex.cbSize, sizeof WNDCLASSEX
    mov     wcex.style, CS_HREDRAW or CS_VREDRAW
    lea     rdi, WndProc
    mov     wcex.lpfnWndProc, rdi
    mov     wcex.cbClsExtra, NULL
    mov     wcex.cbWndExtra, NULL
    mov     rax, hInst
    mov     wcex.hInstance, rax
    mov     rax, hIcon
    mov     wcex.hIcon, rax
    mov     rbx, hCursor
    mov     wcex.hCursor, rbx
    mov     wcex.hbrBackground, COLOR_WINDOW + 1
    lea     rdi, MenuName
    mov     wcex.lpszMenuName, rdi
    lea     rdi, ClassName
    mov     wcex.lpszClassName, rdi
    mov     wcex.hIconSm, rax
    invoke  RegisterClassEx, addr wcex
    
    invoke  CreateWindowEx, \
                    WS_EX_CLIENTEDGE, \
                    addr ClassName, \
                    addr AppName, \
                    WS_OVERLAPPEDWINDOW, \
                    CW_USEDEFAULT, \
                    CW_USEDEFAULT, \
                    300, \
                    200, \
                    NULL, \
                    NULL, \
                    hInst, \
                    NULL
    mov     hWnd, rax
    
    invoke  ShowWindow, hWnd, SW_SHOWNORMAL
    invoke  UpdateWindow, hWnd
    
    .while TRUE
        invoke  GetMessage, addr msg, NULL, 0, 0
        .if (rax == 0)
            .break
        .endif
        invoke  TranslateMessage, addr msg
        invoke  DispatchMessage, addr msg
    .endw
    
    mov     rax, msg.wParam
    
    ret

WinMain endp

WndProc proc hWnd:HWND, uMsg:UINT, wParam:WPARAM, lParam:LPARAM
    
    .if uMsg==WM_CREATE
        invoke GetMenu, hWnd
        mov hMenu, rax
        
        mov ofn.lStructSize, sizeof ofn
        mov rax, hWnd
        mov ofn.hwndOwner, rax
        mov rax, hInstance
        mov ofn.hInstance, rax
        lea rdi, FilterStr
        mov ofn.lpstrFilter, rdi
        lea rdi, buffer
        mov ofn.lpstrFile, rdi
        mov ofn.nMaxFile, MAXSIZE
    .elseif uMsg==WM_COMMAND
        mov rax, wParam
        .if lParam==0
            .if ax==IDM_OPEN
                mov ofn.Flags,  OFN_FILEMUSTEXIST or OFN_PATHMUSTEXIST or \
                                OFN_LONGNAMES or OFN_EXPLORER or OFN_HIDEREADONLY
                invoke GetOpenFileName, addr ofn
                .if rax==TRUE
                    invoke  CreateFile,addr buffer, \
                                    GENERIC_READ, \
                                    NULL, \
                                    NULL, \
                                    OPEN_EXISTING, \
                                    FILE_ATTRIBUTE_ARCHIVE, \
                                    NULL
                    mov hFileRead, rax
                    invoke  CreateFileMapping, hFileRead, NULL, PAGE_READONLY, NULL, NULL, NULL
                    mov hMapFile, rax
                    lea rax, buffer
                    movzx edx, ofn.nFileOffset
                    add rax, rdx
                    
                    invoke SetWindowText, hWnd, rax
                    invoke EnableMenuItem, hMenu, IDM_OPEN, MF_GRAYED
                    invoke EnableMenuItem, hMenu, IDM_SAVE, MF_ENABLED
                .endif
            .elseif ax==IDM_SAVE
                mov ofn.Flags, OFN_LONGNAMES or OFN_EXPLORER or OFN_HIDEREADONLY
                invoke GetSaveFileName, addr ofn
                .if rax==TRUE
                    invoke  CreateFile, addr buffer, \
                                    GENERIC_READ or GENERIC_WRITE, \
                                    FILE_SHARE_READ or FILE_SHARE_WRITE, \
                                    NULL, \
                                    CREATE_NEW, \
                                    FILE_ATTRIBUTE_ARCHIVE, \
                                    NULL
                    mov hFileWrite, rax
                    invoke MapViewOfFile, hMapFile, FILE_MAP_READ, 0, 0, 0
                    mov pMemory, rax
                    invoke GetFileSize, hFileRead, NULL
                    invoke WriteFile, hFileWrite, pMemory, rax, addr SizeWritten, NULL
                    invoke CloseMapFile
                    invoke CloseHandle, hFileWrite
                    invoke SetWindowText, hWnd, addr AppName
                    invoke EnableMenuItem, hMenu, IDM_OPEN, MF_ENABLED
                    invoke EnableMenuItem, hMenu, IDM_SAVE, MF_GRAYED
                .endif
            .else
                invoke DestroyWindow, hWnd
            .endif
        .endif
    .elseif uMsg==WM_DESTROY
        .if hMapFile {} 0
            invoke CloseMapFile
        .endif
        invoke  PostQuitMessage, NULL
    .else
        invoke  DefWindowProc, hWnd, uMsg, wParam, lParam
        ret
    .endif
    xor eax, eax
    ret

WndProc endp

CloseMapFile proc
    
    invoke  CloseHandle, hMapFile
    mov     hMapFile, 0
    invoke  CloseHandle, hFileRead
    ret

CloseMapFile endp

end
